File: 19.remove-nth-node-from-end-of-list.cpp
/*
 * @lc app=leetcode id=19 lang=cpp
 *
 * [19] Remove Nth Node From End of List
 */
struct
    ListNode;
// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    ListNode *removeNthFromEnd(ListNode *head, int n)
    {
        ListNode *preDel = head;
        ListNode *cur = head;
        for (int i = n; i != 0; i--)
        {
            cur = cur->next;
        }
        //if cur is nullptr, delete the first Node;
        if (cur == nullptr)
        {
            auto res = head->next;
            delete (head);
            return res;
        }
        else
        {
            cur = cur->next;
        }

        while (cur != nullptr)
        {
            cur = cur->next;
            preDel = preDel->next;
        }
        auto delNode = preDel->next;

        preDel->next = delNode->next;
        //delete (delNode);
        return head;
    }
};
// @lc code=end

File: py/19.remove-nth-node-from-end-of-list.py
#
# @lc app=leetcode id=19 lang=python3
#
# [19] Remove Nth Node From End of List
#
from typing import*


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next


class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(next=head)
        slow = fast = dummy
        while fast and n:
            fast = fast.next
            n -= 1
        if n:
            return dummy.next
        while fast.next:
            slow = slow.next
            fast = fast.next
        slow.next = slow.next.next
        return dummy.next

# @lc code=end
